exec/java-exec/src/test/java/org/apache/drill/test/TestGracefulShutdown.java

```diff
@@ -0,0 +1,248 @@
+    props.putAll(WEBSERVER_CONFIGURATION);
+    builder.configBuilder.configProps(props);
+    return builder;
+  }
+
+
+  /*
+  Start multiple drillbits and then shutdown a drillbit. Query the online
+  endpoints and check if the drillbit still exists.
+   */
+  @Test
+  public void testOnlineEndPoints() throws  Exception {
+
+    String[] drillbits = {"db1" ,"db2","db3", "db4", "db5", "db6"};
+    ClusterFixtureBuilder builder = ClusterFixture.builder().withBits(drillbits).withLocalZk();
+
+
+    try ( ClusterFixture cluster = builder.build();
+          ClientFixture client = cluster.clientFixture()) {
+
+      Drillbit drillbit = cluster.drillbit("db2");
+      DrillbitEndpoint drillbitEndpoint =  drillbit.getRegistrationHandle().getEndPoint();
+      int grace_period = drillbit.getContext().getConfig().getInt("drill.exec.grace_period");
+      new Thread(new Runnable() {
+        public void run() {
+          try {
+            cluster.closeDrillbit("db2");
+          } catch (Exception e) {
+            e.printStackTrace();
+          }
+        }
+      }).start();
+      //wait for graceperiod
+      Thread.sleep(grace_period);
+      Collection<DrillbitEndpoint> drillbitEndpoints = cluster.drillbit().getContext()
+              .getClusterCoordinator()
+              .getOnlineEndPoints();
+      Assert.assertFalse(drillbitEndpoints.contains(drillbitEndpoint));
+    }
+  }
+  /*
+    Test if the drillbit transitions from ONLINE state when a shutdown
+    request is initiated
+   */
+  @Test
+  public void testStateChange() throws  Exception {
+
+    String[] drillbits = {"db1" ,"db2", "db3", "db4", "db5", "db6"};
+    ClusterFixtureBuilder builder = ClusterFixture.builder().withBits(drillbits).withLocalZk();
+
+    try ( ClusterFixture cluster = builder.build();
+          ClientFixture client = cluster.clientFixture()) {
+      Drillbit drillbit = cluster.drillbit("db2");
+      int grace_period = drillbit.getContext().getConfig().getInt("drill.exec.grace_period");
+      DrillbitEndpoint drillbitEndpoint =  drillbit.getRegistrationHandle().getEndPoint();
+      new Thread(new Runnable() {
+        public void run() {
+          try {
+            cluster.closeDrillbit("db2");
+          } catch (Exception e) {
+            e.printStackTrace();
+          }
+        }
+      }).start();
+      Thread.sleep(grace_period);
+      Collection<DrillbitEndpoint> drillbitEndpoints = cluster.drillbit().getContext()
+              .getClusterCoordinator()
+              .getAvailableEndpoints();
+      for (DrillbitEndpoint dbEndpoint : drillbitEndpoints) {
+        if(drillbitEndpoint.getAddress().equals(dbEndpoint.getAddress()) && drillbitEndpoint.getUserPort() == dbEndpoint.getUserPort()) {
+          assertNotEquals(dbEndpoint.getState(),DrillbitEndpoint.State.ONLINE);
+        }
+      }
+    }
+  }
+
+  /*
+   Test shutdown through RestApi
+   */
+  @Test
+  public void testRestApi() throws Exception {
```

Comment:
@ilooner provided a REST test fixture in a recent PR. Should this test use that?